nemo/collections/speechlm2/models/duplex_ear_tts.py

```diff
@@ -0,0 +1,1522 @@
+        )
+    else:
+        return torch.where(torch.isin(speech_codes, control_codes), speech_codes[:, :1], speech_codes)
+
+
+def get_mask_from_lengths(
+    lengths: torch.Tensor = None, x: torch.Tensor = None, pad_to_factor: int = None
+) -> torch.Tensor:
+    """Constructs binary mask from a 1D torch tensor of input lengths
+    Args:
+        lengths: torch.tensor (torch.tensor): 1D tensor with lengths
+        x: torch.tensor = tensor to be used on, last dimension is for mask
+    Returns:
+        mask (torch.tensor): num_sequences x max_length binary tensor
+    """
+    if lengths is None:
+        assert x is not None
+        return torch.ones(x.shape[-1], dtype=torch.bool, device=x.device)
+    else:
+        if x is None:
+            max_len = torch.max(lengths)
+        else:
+            max_len = x.shape[-1]
+
+    if pad_to_factor is not None:
+        with fp32_precision():
+            max_len = torch.ceil(max_len / pad_to_factor) * pad_to_factor
+
+    ids = torch.arange(0, max_len, device=lengths.device, dtype=lengths.dtype)
+    mask = ids < lengths.unsqueeze(1)
+    return mask
+
+
+def setup_rvq_audio_codec(model):
+    """
+    Sets up an ``AudioCodecModel``, initializing it from pretrained weights.
+    The result is assigned to ``model.audio_codec`` attribute.
+
+    Includes a workaround for PTL auto-downcasting the codec model to bf16 with bf16-true precision.
+    """
+    if hasattr(model, "audio_codec") and next(model.audio_codec.parameters()).dtype == model.audio_codec_run_dtype:
+        return  # skip if already set up and has the right dtype
+
+    with ensures_target_precision(model.audio_codec_run_dtype):
+        if model.cfg.get("pretrained_ae_dir", None):
+            model.audio_codec = (
+                RVQVAEModel.from_pretrained(
+                    model.cfg.pretrained_ae_dir,
+                    cfg=DictConfig(model.cfg.codec_config) if model.cfg.get("codec_config", None) else None,
+                    strict=False,
+                )
+                .eval()
+                .to(model.device)
+            )
+        else:
+            # init codec from config
+            model.audio_codec = RVQVAEModel(DictConfig(model.cfg.codec_config))
+
+    for p in model.audio_codec.parameters():
+        p.requires_grad = False
+
+
+def setup_audio_codec(self):
+    setup_rvq_audio_codec(self)
+    assert callable(self.tts_model.set_rvq_embs)
+    self.tts_model.set_rvq_embs(torch.stack([x.detach() for x in self.audio_codec.prvq.mus_list], 0))
+    self.tts_model.rvq_embs = self.tts_model.rvq_embs.to(next(self.tts_model.parameters()).dtype)
+    # compute target fps
+    self.target_fps = self.target_sample_rate / self.audio_codec.config.wav_to_token_ratio
+    self.target_samples_per_frame = self.audio_codec.config.wav_to_token_ratio
+
+
+def rescale_state_dict(state_dict, target_std=0.02, first_n_layers=None, layer_prefix="tts_model.backbone.layers."):
```

Comment:
Do we still use this? If so, can we replace prints with logging. If not, let's remove